src/StoreBase.ts

```diff
@@ -60,178 +56,211 @@ export abstract class StoreBase {
         assert.ok(this._triggerBlockCount >= 0, 'Over-popped trigger blocks!');
 
         if (this._triggerBlockCount === 0) {
-            // Go through the list of stores awaiting resolution and resolve them all
-            const awaitingList = this._triggerBlockedStoreList;
-            this._triggerBlockedStoreList = [];
-            _.forEach(awaitingList, store => {
-                store._resolveThrottledCallbacks();
-            });
+            StoreBase._resolveCallbacks();
         }
     }
 
     static setThrottleStatus(enabled: boolean) {
         this._bypassThrottle = !enabled;
+        
+        StoreBase._resolveCallbacks();
+    }
 
-        // If we're going to bypass the throttle, trigger all pending stores now
-        if (this._bypassThrottle) {
-            let pendingThrottledStore = this._pendingThrottledStores.shift();
-            while (!!pendingThrottledStore) {
-                pendingThrottledStore._resolveThrottledCallbacks();
-                pendingThrottledStore = this._pendingThrottledStores.shift();
-            }
-        }
+    constructor(private _throttleMs?: number, private _bypassTriggerBlocks = false) {
     }
 
-    constructor(throttleMs: number = 0, bypassTriggerBans = false) {
-        this._throttleMs = throttleMs;
-        this._bypassTriggerBlocks = bypassTriggerBans;
+    private static _updateExistingMeta(meta: CallbackMetadata | undefined, throttledUntil: number|undefined, bypassBlock: boolean) {
+        if (!meta) {
+            return;
+        }
+        // Update throttling value to me min of exiting and new value
+        if (throttledUntil && meta.throttledUntil) {
+            meta.throttledUntil = Math.min(meta.throttledUntil, throttledUntil);
+        }
+
+        if (!throttledUntil) {
+            meta.throttledUntil = undefined;
+        }
+
+        if (bypassBlock) {
+            meta.bypassBlock = true;
+        }
     }
 
     // If you trigger a specific set of keys, then it will only trigger that specific set of callbacks (and subscriptions marked
     // as "All" keyed).  If the key is all, it will trigger all callbacks.
     protected trigger(keyOrKeys?: string|number|(string|number)[]) {
-        let keys: string[]|undefined;
-
-        // trigger(0) is valid, ensure that we catch this case
-        if (keyOrKeys || _.isNumber(keyOrKeys)) {
-            keys = _.map(_.isArray(keyOrKeys) ? keyOrKeys : [keyOrKeys], key => _.isNumber(key) ? key.toString() : key);
+        // If we're throttling, save execution time
+        let throttledUntil: number | undefined;
+        let throttleMs: number | undefined;
+        if (this._throttleMs !== undefined) {
+            throttleMs = this._throttleMs;
+        } else {
+            // If the store doens't define any throttling, pick up the default
+            throttleMs = Options.defaultThrottleMs;
+        }
+        if (throttleMs) {
+            if (!this._throttleData) {
```

Comment:
It might be a bit nicer to read if you move `_throttleData` down to a separate `if`.
```ts
let throttledUntil: number | undefined;
if (throttleMs) {
    throttleUntil = Date.now() + throttleMs;
}
if (throttleUntil && !this._throttleData) {
    this._throttleData = {~};
}
```

Reply:
I actually find the existing more readable